### PR TITLE
NeVa token fusion

### PR DESCRIPTION
**Pull request by paul-gibbons**
# What does this PR do ?

Adding in token fusion via mlp_downsample adapter from VILA paper.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by Slyne — COMMENTED** [review]
LGTM. 
No impact on my side.

File: nemo/collections/nlp/modules/common/text_generation_strategy.py [hunk 1]
```diff
@@ -545,15 +544,21 @@ def __init__(self, model):
             sep_image_conv_front=self.data_cfg.sep_image_conv_front,
             conv_template=self.data_cfg.get("conv_template", "nvgpt"),
             model_type=self.cfg.mm_cfg.llm.get("model_type", "nvgpt"),
-            image_token_len=self.data_cfg.image_token_len,
-            image_folder=self.data_cfg.image_folder,
+            patch_dim=self.cfg.mm_cfg.vision_encoder.patch_dim,
+            crop_size=self.cfg.mm_cfg.vision_encoder.get("crop_size", (224, 224)),
```
Comment (yaoyu-33) [inline nemo/collections/nlp/modules/common/text_generation_strategy.py:548]: Can you change the logic here, and default this to None. If this is none, and get crop size from HF processor?

File: nemo/collections/multimodal/data/neva/neva_dataset.py [hunk 2]
```diff
@@ -265,6 +263,9 @@ def preprocess_multimodal(sources: dict, multimodal_cfg: dict, cur_token_len: in
     if media_type == 'video':
         num_patches *= multimodal_cfg['num_frames']
 
+    if multimodal_cfg['mm_mlp_adapter_type'] == 'mlp_downsample':
+        num_patches //= 4
```
Comment (yaoyu-33) [inline nemo/collections/multimodal/data/neva/neva_dataset.py:267]: did you handle padding somewhere else? for odd number dimension

File: nemo/collections/multimodal/data/neva/neva_dataset.py [hunk 3]
```diff
@@ -1212,8 +1217,9 @@ def make_supervised_data_module(tokenizer, model_cfg) -> Dict:
             sep_image_conv_front=data_cfg.sep_image_conv_front,
             model_type=mm_cfg.llm.get("model_type", "nvgpt"),
             conv_template=data_cfg.get("conv_template", "nvgpt"),
+            patch_dim=model_cfg.mm_cfg.vision_encoder.patch_dim,
             crop_size=crop_size,
-            image_token_len=data_cfg.image_token_len,
+            image_token_len=data_cfg.get('image_token_len', None),
```
Comment (yaoyu-33) [inline nemo/collections/multimodal/data/neva/neva_dataset.py:1222]: we should remove this for good

File: nemo/collections/nlp/modules/common/text_generation_strategy.py [hunk 4]
```diff
@@ -545,15 +544,30 @@ def __init__(self, model):
...
+        if self.multimodal_cfg['crop_size'] is None:
+            image_processor = CLIPImageProcessor.from_pretrained(
+                self.cfg.mm_cfg.vision_encoder.from_pretrained, torch_dtype=torch.bfloat16
+            )
+            self.multimodal_cfg['crop_size'] = (
+                image_processor.crop_size['height'],
+                image_processor.crop_size['width'],
+            )
+
+        self.num_media_latents = (self.multimodal_cfg['crop_size'][0] // self.multimodal_cfg['patch_dim']) * (
```
Comment (yaoyu-33) [inline nemo/collections/nlp/modules/common/text_generation_strategy.py:569]: consider odd number crop size

**Review by yaoyu-33 — APPROVED** [review]

**Pull request merged**